Saving

graberyanis opened the pull request:
- added file manager for opening and reading from/loading to files
- main: 
- uses file manager to open file, saves networks and generation counter into dictionary after ending the simulation and loads dictionary to json file.
- layer:
- layers can be given inputs for weights and biases, for using save data, if none are given, creates random weights/biases
- network:
- networks can be given weights/biases as inputs which will be used to create the single layers
- population:
- population can be given a list with weights and biases of another generation which will be used to create a new population
- simulation:
- uses save dictionary for creating the population and setting the generation counter, simulation will now only stop after the current generation has ended (to avoid distortion in data) (will later also change, so population will only be reset if the simulation is running)

graberyanis committed:
increased mutations and changed network sizes to work with 0 eye size

graberyanis committed:
allowed for weights and biases to be given to the layer
instead of creating random new ones

graberyanis committed:
if weights and biases are given, uses them for the layers

graberyanis committed:
added class for loading from and dumping to json files

graberyanis committed:
simulation now stops after the current generation has finished

graberyanis committed:
if a save file dictionary is given, uses it to recreate the networks from the save

graberyanis committed:
creates a dictionary with all the networks values and the generation counter value

webermael commented on src/main.py:
```diff
@@ -6,10 +6,24 @@
 
 def main():
-    simulation = Simulation()
+    FM = FileManager()
```
Would you mind making this lowercase? Just to keep everything _uniform_ ...

webermael reviewed (approved):
Verry clean code (and everything still works 🙃)

graberyanis committed:
only creates new gen while still running
already loads some variables from save file

graberyanis committed:
updates the variables of the current save, which are changed by the simulation

graberyanis committed:
changed input dictionary hierarchy

graberyanis committed:
now updates the output dictionary through file manager

graberyanis committed:
no more config!

graberyanis committed:
Using input file instead of config for everything
+ plus some code cleanup and cosmetics :)

graberyanis committed:
Merge branch 'main' into saving

graberyanis commented:
The config file has been made redundant by the almighty input_dict, who came from the far land named: template.json
Also fire is now really burning with a cool new color changing effect!
Fire won't set fire on fire anymore
Bigger better and blue organisms for more enjoyement (also now equipped with radars to see where they are :D)

webermael commented on src/evolution/mutation.py:
```diff
@@ -1,6 +1,6 @@
-    if random.random() <= config.mutation_rate:
-        return value + random.uniform(config.mutation_max_difference, -config.mutation_max_difference)
+def mutation(inputs, value):
+    if random.random() <= inputs["chance"]:
```
Shouldn't it be called "probability" instead of "chance"?

graberyanis replied:
good point

webermael commented on src/main.py:
```diff
@@ -1,15 +1,21 @@
-#import pygame
-#from config import Config
-#from world.world import World
-#from organisms.population import Population
-#from utils.visualisation import (render_world, screen_init, screen_update)
-#from evolution.reproduction import reproduce
-#from evolution.mutation import mutate
```
at last

webermael commented on src/organisms/organism.py:
```diff
@@ -29,37 +29,37 @@ def __init__(self, config, x_pos, y_pos, nn = False):
+                    self.score += inputs["population"]["interactions"]["move_bonus"]
         else:
-            self.lifetime -= config.border_damage
+            self.lifetime -= inputs["population"]["interactions"]["border_damage"]
```
Just to throw it in there: do we still want to take damage at the border?
I think we once talked about "wrapping around", so the organisms can move "freely"...
-> What did we decide there?

graberyanis replied:
I would leave it for now, maybe change it after having the branch and the interface together...

webermael commented on src/organisms/sensors.py:
```diff
@@ -4,15 +4,15 @@ def __init__(self, size):
         self.translation = {"danger" : 0, "empty" : 0.4, "food" : 1}
         self.org_translation = {"danger" : 0, "empty" : 0.6, "food" : 0.8}
```
What are we doing about that now? I don't think we really came to a conclusion on this.

graberyanis replied:
I feel like it worked out, the biggest issue i feel like is them not comprehending the edge of the world

webermael reviewed (approved):
Danger looks _fire_

graberyanis committed:
added template file and loaded it from inside the project

graberyanis committed:
changed "chance" to "probability"

graberyanis committed:
world now generates strings, then replaces them with object
as empty tiles are replaced by food and danger, this way there won't be unnecessary empty tile objects

graberyanis committed:
Merge branch 'saving' of https://github.com/webermael/NAWIMAT-Projektarbeit into saving
